src/Symfony/Component/AutoMapper/AutoMapper.php

```diff
@@ -0,0 +1,273 @@
+
+        if (\array_key_exists($className, $this->mapperRegistry)) {
+            return $this->mapperRegistry[$className];
+        }
+
+        if (!class_exists($className)) {
+            $this->classLoader->loadClass($metadata);
+        }
+
+        $this->mapperRegistry[$className] = new $className();
+        $this->mapperRegistry[$className]->injectMappers($this);
+
+        foreach ($metadata->getCallbacks() as $property => $callback) {
+            $this->mapperRegistry[$className]->addCallback($property, $callback);
+        }
+
+        return $this->mapperRegistry[$className];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasMapper(string $source, string $target): bool
+    {
+        return null !== $this->getMetadata($source, $target);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function map($sourceData, $targetData, Context $context = null)
+    {
+        $source = null;
+        $target = null;
+
+        if (null === $sourceData) {
+            return null;
+        }
+
+        if (\is_object($sourceData)) {
+            $source = \get_class($sourceData);
+        }
+
+        if (\is_array($sourceData)) {
+            $source = 'array';
+        }
+
+        if (null === $source) {
+            throw new NoMappingFoundException('Cannot map this value, source is neither an object or an array.');
+        }
+
+        if (null === $context) {
+            $context = new Context();
+        }
+
+        if (\is_object($targetData)) {
+            $target = \get_class($targetData);
+            $context->setObjectToPopulate($targetData);
+        }
+
+        if (\is_array($targetData)) {
+            $target = 'array';
+            $context->setObjectToPopulate($targetData);
+        }
+
+        if (\is_string($targetData)) {
+            $target = $targetData;
+        }
+
+        if (null === $target) {
+            throw new NoMappingFoundException('Cannot map this value, target is neither an object or an array.');
+        }
+
+        if ('array' === $source && 'array' === $target) {
+            throw new NoMappingFoundException('Cannot map this value, both source and target are array.');
+        }
+
+        return $this->getMapper($source, $target)->map($sourceData, $context);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getMetadata(string $source, string $target): ?MapperGeneratorMetadataInterface
+    {
+        if (!\array_key_exists($source, $this->metadatas) || !\array_key_exists($target, $this->metadatas[$source])) {
+            if (null === $this->mapperConfigurationFactory) {
+                return null;
+            }
+
+            $this->register($this->mapperConfigurationFactory->create($this, $source, $target));
+        }
+
+        return $this->metadatas[$source][$target];
+    }
+
+    /**
+     * Use this for test, benchmark and fast prototyping.
+     *
+     * @internal
+     */
+    public static function create(
```

Comment:
This kind of methods are very useful when using the component without the "full stack" framework. We lack such method for the Serializer component, it would be nice to have it not marked `@internal`, but it can be done later.